cardano-crypto-class/src/Cardano/Crypto/Libsodium/Memory/Internal.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+module Cardano.Crypto.Libsodium.Memory.Internal (
+  -- * High-level memory management
+  MLockedForeignPtr (..),
+  withMLockedForeignPtr,
+  allocMLockedForeignPtr,
+  finalizeMLockedForeignPtr,
+  traceMLockedForeignPtr,
+  -- * Low-level memory function
+  sodiumMalloc,
+  sodiumFree,
+) where
+
+import Control.Monad (when)
+import Data.Coerce (coerce)
+import Foreign.C.Error (errnoToIOError, getErrno)
+import Foreign.C.Types (CSize (..))
+import Foreign.ForeignPtr (ForeignPtr, newForeignPtr, withForeignPtr, finalizeForeignPtr)
+import Foreign.Ptr (Ptr, nullPtr)
+import Foreign.Storable (Storable (alignment, sizeOf, peek))
+import GHC.IO.Exception (ioException)
+
+import Cardano.Prelude (NoUnexpectedThunks, OnlyCheckIsWHNF (..))
+import Cardano.Crypto.Libsodium.C
+
+-- | Foreign pointer to securely allocated memory.
+newtype MLockedForeignPtr a = SFP { _unwrapMLockedForeignPtr :: ForeignPtr a }
+  deriving NoUnexpectedThunks via OnlyCheckIsWHNF "MLockedForeignPtr" (MLockedForeignPtr a)
+
+withMLockedForeignPtr :: forall a b. MLockedForeignPtr a -> (Ptr a -> IO b) -> IO b
+withMLockedForeignPtr = coerce (withForeignPtr @a @b)
+
+finalizeMLockedForeignPtr :: forall a. MLockedForeignPtr a -> IO ()
+finalizeMLockedForeignPtr = coerce (finalizeForeignPtr @a)
+
+traceMLockedForeignPtr :: (Storable a, Show a) => MLockedForeignPtr a -> IO ()
+traceMLockedForeignPtr fptr = withMLockedForeignPtr fptr $ \ptr -> do
+    a <- peek ptr
+    print a
+
+{-# DEPRECATED traceMLockedForeignPtr "Don't leave traceMLockedForeignPtr in production" #-}
```

Comment:
Ah, deprecation is for this? Makes sense..

Reply:
Yes, makes `-Werror` useful and not only annoying ;)